interrupt trajectory visualization on arrival of new display trajectory

rhaschke opened the pull request:
A failed planning request often results in a rather long display trajectory. However, the next trajectory was only displayed, when the current one has finished.
The first patch immediately displays this trajectory when the planner was invoked from rviz - for responsive user interaction.
The second patch introduces a new display property to allow the user to switch on/off this interruption behavior in all other cases too.

rhaschke committed:
immediately show trajectory after planning (interrupting current display)

rhaschke committed:
new GUI property to allow immediate interruption of displayed trajectory

rethink-imcmahon commented:
I'm thrilled you've addressed this issue, @rhaschke! Failed planning attempts meant wasting 20 seconds while the robot flails wildly in RViz...

Not having worked on the internals of RViz or MoveIt before, I will defer making a technical review to others that have more experience with them. But with just a cursory glance, I think these changes look good. I very much want to see this merged.

rhaschke committed:
cleanup TrajectoryVisualization::update

simplified code to switch to new trajectory / start over animation in loop mode

rhaschke committed:
fixed race condition for trajectory-display interruption

- TrajectoryVisualization::update() switches to new trajectory
  automatically when it has finished displaying the old one
- TrajectoryVisualization::interruptCurrentDisplay() might interrupt
  this newly started trajectory

consequences:
- protect switching of trajectory with mutex
- interrupt only if trajectory display progressed past first waypoint
- removed obsolete signal timeToShowNewTrail:
  update() automatically switches and updates trail in sync

rhaschke committed:
correctly setAlpha for new trail

rhaschke committed:
added missing initialization

rhaschke committed:
fixup! cleanup TrajectoryVisualization::update

only enter visualization loop when displaying_trajectory_message_ is defined

sachinchitta commented:
Tested this - nice change!

davetcoleman commented:
this is great, thanks @rhaschke
